testutil/chan.go

```diff
@@ -55,3 +55,61 @@ func RequireSend[A any](ctx context.Context, t testing.TB, c chan<- A, a A) {
 		// OK!
 	}
 }
+
+// SoftTryReceive will attempt to receive a value from the chan and return it. If
+// the context expires before a value can be received, it will mark the test as
+// failed but continue execution. If the channel is closed, the zero value of the
+// channel type will be returned.
+// The second return value indicates whether the receive was successful. In
+// particular, if the channel is closed, the second return value will be true.
+//
+// Safety: can be called from any goroutine.
+func SoftTryReceive[A any](ctx context.Context, t testing.TB, c <-chan A) (A, bool) {
+	t.Helper()
+	select {
+	case <-ctx.Done():
+		t.Error("timeout")
+		var a A
+		return a, false
+	case a := <-c:
+		return a, true
+	}
+}
+
+// AssertReceive will receive a value from the chan and return it. If the
+// context expires or the channel is closed before a value can be received,
+// it will mark the test as failed but continue execution.
+// The second return value indicates whether the receive was successful.
+//
+// Safety: can be called from any goroutine.
+func AssertReceive[A any](ctx context.Context, t testing.TB, c <-chan A) (A, bool) {
```

Comment:
Nit: could we abstract the logic and either use `t.Fatal` or `t.Error` in either case? That seems to be the only delta between the implementations.

Reply:
There's also the extra return value. I suppose we could create internal helper functions which would be called with a bool param by the Require/Assert variants. Not sure if it's worth it here - what do you think?